File: backend/app/schemas/analytics.py
from typing import Any

from app.schemas.common import ORMModel


class AnalyticsAnswer(ORMModel):
    value: Any
    count: int


class QuestionAnalytics(ORMModel):
    question_id: int
    question_title: str
    question_type: str
    response_count: int
    answers: list[AnalyticsAnswer]


class Analytics(ORMModel):
    form_id: int
    total_responses: int
    started_responses: int
    completed_responses: int
    average_completion_time_seconds: int | float | None
    questions: list[QuestionAnalytics]

File: frontend/src/types/analytics.ts
import type { QuestionType } from "./question";

export interface ChoiceAnalyticsCount {
  label: string;
  count: number;
  percentage: number;
}

export interface ChoiceAnalyticsSummary {
  counts: ChoiceAnalyticsCount[];
}

export interface RatingDistributionItem {
  value: number;
  count: number;
}

export interface RatingAnalyticsSummary {
  average: number | null;
  distribution: RatingDistributionItem[];
}

export interface YesNoAnalyticsSummary {
  yes: number;
  no: number;
}

export type QuestionAnalyticsSummary =
  | ChoiceAnalyticsSummary
  | RatingAnalyticsSummary
  | YesNoAnalyticsSummary
  | null;

export interface QuestionAnalytics {
  question_id: number;
  title: string;
  type: QuestionType;
  answered_count: number;
  skipped_count: number;
  summary: QuestionAnalyticsSummary;
}

export interface FormAnalytics {
  form_id: number;
  form_title: string;
  total_responses: number;
  started_responses: number;
  completed_responses: number;
  average_completion_time_seconds: number | null;
  questions: QuestionAnalytics[];
}
